cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -217,8 +214,11 @@ protected void shutDown() throws Exception {
     for (ApplicationSpecification appSpec : store.getAllApplications(namespace)) {
       appIds.add(namespace.app(appSpec.getName(), appSpec.getAppVersion()));
     }
+    Set<? extends EntityId> visibleAppIds = authorizationEnforcer.isVisible(appIds,
+                                                                            authenticationContext.getPrincipal());
 
-    for (ApplicationId appId : appIds) {
+    for (EntityId entityId : visibleAppIds) {
+      ApplicationId appId = (ApplicationId) entityId;
       ApplicationSpecification appSpec = store.getApplication(appId);
```

Comment:
If we store the app specs that we get from `store.getAllApplications(namespace)`, then we don't need to make calls to store to fetch each app spec again.